### PR TITLE
Faster post-processing with `device=mps`

### PR DESCRIPTION
**Pull request by Y-T-G**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Performance and stability improvements for validation, especially on Apple MPS (Mac) devices, plus a minor device-placement fix to prevent errors. 🚀

### 📊 Key Changes
- Optimize post-processing on MPS: move tensors to CPU before postprocess when `device.type == "mps"` for faster execution.
- Fix device mismatch in detection validation: compute bbox scaling using `bbox.device` instead of `self.device` to avoid tensor device conflicts.

### 🎯 Purpose & Impact
- Faster validation on Macs with MPS by leveraging quicker CPU post-processing. ⚡
- Increased robustness: prevents device mismatch errors during evaluation. 🛠️
- No API changes; transparent improvements for users running validation/evaluation workflows. ✅

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline reviews. An Ultralytics engineer will also assist you here shortly.

- ✅ **Define a Purpose**: Your summary and intent around MPS post-processing and the device-placement fix are clear. Please ensure your PR description links to any [relevant issues](https://github.com/ultralytics/ultralytics/issues) (performance or device mismatch), and that your commit messages are concise and follow project conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with `ultralytics/ultralytics` `main`. If it's behind, update it with the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address them.
- ✅ **Update Documentation**: If applicable, update [documentation](https://docs.ultralytics.com/) for behavior changes (e.g., notes on MPS validation performance or device handling in validators).
- ✅ **Add Tests**: If possible, add or update tests to cover the MPS/CPU post-processing path and the bbox scaling device fix. Confirm that all tests pass.
- ✅ **Sign the CLA**: If this is your first Ultralytics PR, please sign our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) by commenting: "I have read the CLA Document and I sign the CLA".
- ✅ **Minimize Changes**: Keep the changeset focused on MPS post-processing and the device mismatch fix only. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Optional but helpful:
- 📈 Please include brief before/after validation timings on macOS/MPS (model, imgsz, batch size, dataset split) to quantify the speedup.
- 🧪 If the device mismatch fix addresses a specific issue, link it in the PR description and add a short note on how you validated the fix.

For more guidance, see our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Thanks for improving the validation experience on MPS and robustness of device placement! ⚡🛠️

**Comment by Y-T-G** [comment]
@glenn-jocher I don't know if this is a bug or something wrong with MPS, but post-processing is magnitudes faster on CPU than MPS for me:

**main**
```bash
yolo val data=coco128.yaml device=mps imgsz=160 plots=False
Ultralytics 8.3.189 🚀 Python-3.12.0 torch-2.8.0 MPS (Apple M4 Pro)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2085.8±716.6 MB/s, size: 58.1 KB)
val: Scanning /Users/toxite/Downloads/GitHub/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 128/128 31580641val: Scanning /Users/toxite/Downloads/GitHub/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 128/128 4294967.3it/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 8/8 2.3it/s 3.5s
                   all        128        929       0.53      0.239      0.254      0.179
                person         61        254      0.507      0.333      0.369      0.229
               bicycle          3          6          0          0     0.0194     0.0155
                   car         12         46      0.278     0.0652     0.0414     0.0209
            motorcycle          4          5      0.418        0.6      0.508      0.382
              airplane          5          6      0.735        0.5      0.599      0.509
                   bus          5          7      0.671      0.286       0.53      0.451
                 train          3          3      0.184      0.333      0.289     0.0896
                 truck          5         12          0          0        0.1     0.0279
                  boat          2          6      0.536      0.333      0.451      0.303
         traffic light          4         14          1          0     0.0295     0.0208
             stop sign          2          2          1          0          0          0
                 bench          5          9       0.29      0.111      0.125     0.0857
                  bird          2         16       0.28      0.188      0.156     0.0897
                   cat          4          4       0.45          1      0.687      0.471
                   dog          9          9      0.477      0.409       0.44      0.365
                 horse          1          2     0.0908        0.5      0.262     0.0587
              elephant          4         17      0.731      0.322      0.356      0.266
                  bear          1          1       0.54          1      0.995      0.995
                 zebra          2          4      0.827          1      0.995      0.819
               giraffe          4          9      0.688      0.778      0.778      0.517
              backpack          4          6          1          0     0.0118    0.00317
              umbrella          4         18      0.746      0.164      0.267      0.126
               handbag          9         19          1          0    0.00472    0.00331
                   tie          6          7      0.427      0.286      0.355      0.265
              suitcase          2          4      0.215       0.25      0.248      0.223
               frisbee          5          5       0.52        0.2      0.329      0.274
                  skis          1          1          0          0          0          0
             snowboard          2          7      0.894      0.571      0.701      0.353
           sports ball          6          6          1          0          0          0
                  kite          2         10          0          0          0          0
          baseball bat          4          4          1          0          0          0
        baseball glove          4          7      0.342      0.143      0.158      0.127
            skateboard          3          5          1          0     0.0254    0.00254
         tennis racket          5          7          0          0     0.0566      0.026
                bottle          6         18      0.363      0.111      0.105     0.0882
            wine glass          5         16      0.518     0.0625     0.0684     0.0547
                   cup         10         36      0.422     0.0425     0.0727     0.0315
                  fork          6          6      0.577      0.167      0.196      0.173
                 knife          7         16      0.254      0.188      0.103     0.0217
                 spoon          5         22          0          0     0.0163    0.00662
                  bowl          9         28      0.624      0.357      0.345      0.242
                banana          1          1          1          0          0          0
              sandwich          2          2      0.217        0.5      0.224      0.195
                orange          1          4          1          0     0.0646     0.0323
              broccoli          4         11      0.502      0.182      0.189      0.162
                carrot          3         24          1          0       0.35      0.213
               hot dog          1          2      0.456          1      0.663      0.497
                 pizza          5          5      0.632        0.8      0.825      0.714
                 donut          2         14          1      0.208      0.723      0.496
                  cake          4          4      0.325       0.75      0.415      0.376
                 chair          9         35       0.15      0.114      0.112     0.0634
                 couch          5          6      0.308      0.333      0.356      0.211
          potted plant          9         14      0.447     0.0714      0.118     0.0897
                   bed          3          3      0.231      0.333       0.18     0.0554
          dining table         10         13      0.437      0.359      0.328      0.225
                toilet          2          2          1          0     0.0437     0.0393
                    tv          2          2      0.291        0.5      0.498      0.448
                laptop          2          3          0          0     0.0253     0.0202
                 mouse          2          2          1          0          0          0
                remote          5          8          1          0      0.142     0.0409
            cell phone          5          8          1          0          0          0
             microwave          3          3      0.534      0.333       0.43      0.193
                  oven          5          5      0.502        0.2       0.21      0.107
                  sink          4          6      0.302      0.167      0.142     0.0791
          refrigerator          5          5      0.609        0.2      0.301      0.229
                  book          6         29          0          0     0.0142    0.00374
                 clock          8          9        0.6      0.444      0.447      0.219
                  vase          2          2     0.0305     0.0916     0.0846     0.0761
              scissors          1          1          1          0          0          0
            teddy bear          6         21      0.472     0.0952      0.305      0.141
            toothbrush          2          5          1          0     0.0481     0.0236
Speed: 0.1ms preprocess, 4.6ms inference, 0.0ms loss, 10.7ms postprocess per image
```

**PR**
```bash
yolo val data=coco128.yaml device=mps imgsz=160  plots=False
Ultralytics 8.3.189 🚀 Python-3.12.0 torch-2.8.0 MPS (Apple M4 Pro)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2130.0±1045.8 MB/s, size: 48.8 KB)
val: Scanning /Users/toxite/Downloads/GitHub/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 128/128 76695844val: Scanning /Users/toxite/Downloads/GitHub/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 128/128 5651272.8it/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 8/8 11.2it/s 0.7s
                   all        128        929       0.53      0.239      0.254      0.179
                person         61        254      0.507      0.333      0.369      0.229
               bicycle          3          6          0          0     0.0194     0.0155
                   car         12         46      0.278     0.0652     0.0414     0.0209
            motorcycle          4          5      0.418        0.6      0.508      0.382
              airplane          5          6      0.735        0.5      0.599      0.509
                   bus          5          7      0.671      0.286       0.53      0.451
                 train          3          3      0.184      0.333      0.289     0.0896
                 truck          5         12          0          0        0.1     0.0279
                  boat          2          6      0.536      0.333      0.451      0.303
         traffic light          4         14          1          0     0.0295     0.0208
             stop sign          2          2          1          0          0          0
                 bench          5          9       0.29      0.111      0.125     0.0857
                  bird          2         16       0.28      0.188      0.156     0.0897
                   cat          4          4       0.45          1      0.687      0.471
                   dog          9          9      0.477      0.409       0.44      0.365
                 horse          1          2     0.0908        0.5      0.262     0.0587
              elephant          4         17      0.731      0.322      0.356      0.266
                  bear          1          1       0.54          1      0.995      0.995
                 zebra          2          4      0.827          1      0.995      0.819
               giraffe          4          9      0.688      0.778      0.778      0.517
              backpack          4          6          1          0     0.0118    0.00317
              umbrella          4         18      0.746      0.164      0.267      0.126
               handbag          9         19          1          0    0.00472    0.00331
                   tie          6          7      0.427      0.286      0.355      0.265
              suitcase          2          4      0.215       0.25      0.248      0.223
               frisbee          5          5       0.52        0.2      0.329      0.274
                  skis          1          1          0          0          0          0
             snowboard          2          7      0.894      0.571      0.701      0.353
           sports ball          6          6          1          0          0          0
                  kite          2         10          0          0          0          0
          baseball bat          4          4          1          0          0          0
        baseball glove          4          7      0.342      0.143      0.158      0.127
            skateboard          3          5          1          0     0.0254    0.00254
         tennis racket          5          7          0          0     0.0566      0.026
                bottle          6         18      0.363      0.111      0.105     0.0882
            wine glass          5         16      0.518     0.0625     0.0684     0.0547
                   cup         10         36      0.422     0.0425     0.0727     0.0315
                  fork          6          6      0.577      0.167      0.196      0.173
                 knife          7         16      0.254      0.188      0.103     0.0217
                 spoon          5         22          0          0     0.0163    0.00662
                  bowl          9         28      0.624      0.357      0.345      0.242
                banana          1          1          1          0          0          0
              sandwich          2          2      0.217        0.5      0.224      0.195
                orange          1          4          1          0     0.0646     0.0323
              broccoli          4         11      0.502      0.182      0.189      0.162
                carrot          3         24          1          0       0.35      0.213
               hot dog          1          2      0.456          1      0.663      0.497
                 pizza          5          5      0.632        0.8      0.825      0.714
                 donut          2         14          1      0.208      0.723      0.496
                  cake          4          4      0.325       0.75      0.415      0.376
                 chair          9         35       0.15      0.114      0.112     0.0634
                 couch          5          6      0.308      0.333      0.356      0.211
          potted plant          9         14      0.447     0.0714      0.118     0.0897
                   bed          3          3      0.231      0.333       0.18     0.0554
          dining table         10         13      0.437      0.359      0.328      0.225
                toilet          2          2          1          0     0.0437     0.0393
                    tv          2          2      0.291        0.5      0.498      0.448
                laptop          2          3          0          0     0.0253     0.0202
                 mouse          2          2          1          0          0          0
                remote          5          8          1          0      0.142     0.0409
            cell phone          5          8          1          0          0          0
             microwave          3          3      0.534      0.333       0.43      0.193
                  oven          5          5      0.502        0.2       0.21      0.107
                  sink          4          6      0.302      0.167      0.142     0.0791
          refrigerator          5          5      0.609        0.2      0.301      0.229
                  book          6         29          0          0     0.0142    0.00374
                 clock          8          9        0.6      0.444      0.447      0.219
                  vase          2          2     0.0305     0.0916     0.0846     0.0761
              scissors          1          1          1          0          0          0
            teddy bear          6         21      0.472     0.0952      0.305      0.141
            toothbrush          2          5          1          0     0.0481     0.0236
Speed: 0.1ms preprocess, 4.2ms inference, 0.0ms loss, 0.2ms postprocess per image
```

I noticed this when training on COCO128 with MPS, the validation took much longer than training, even though training and validation runs on the same images. I tried downgrading PyTorch to 2.7.0 and it was still slower. But I don't see the same slowdown in the CI runners logs for MPS.

**Comment by glenn-jocher** [comment]
@Y-T-G nice improvement! Should we centralize this in the NMS function so Predict gets the same benefit?

**Comment by Y-T-G** [comment]
@glenn-jocher Do you also have slow postprocessing time on your Mac? I am not sure if it's my environment, or is there something wrong with MPS.

I can see people report unusually long postprocessing times too before on Mac M1:
https://github.com/ultralytics/ultralytics/issues/12173#issuecomment-2118396537

> Should we centralize this in the NMS function so Predict gets the same benefit?

Yes, that would be better.

**Comment by Y-T-G** [comment]
@glenn-jocher I think I found the issue. It's due to this large `max_wh` value which when multiplied with class labels to produce offsets results in huge value:
https://github.com/ultralytics/ultralytics/blob/9e372fc12abd7fd2fee5abe9215fa2125bbadf62/ultralytics/utils/nms.py#L49-L49

I came across this issue when implementing the NMS export. I used normalized coordinates to overcome the issue:
https://github.com/ultralytics/ultralytics/blob/9e372fc12abd7fd2fee5abe9215fa2125bbadf62/ultralytics/engine/exporter.py#L1542

And the results were pretty much identical. I think we should do that here. Big numbers are bad in general for computation.

**Comment by Y-T-G** [comment]
`torch` really has issues with MPS on M4. The validation results changes every run.

```
(base) toxite@macos ultralytics % yolo val model=yolo11n.pt device=mps exist_ok plots=False compile=False
WARNING ⚠️ 'data' argument is missing. Using default 'data=coco8.yaml'.
Ultralytics 8.3.198 🚀 Python-3.12.0 torch-2.6.0 MPS (Apple M4 Pro)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2193.3±982.5 MB/s, size: 54.0 KB)
val: Scanning /Users/toxite/Downloads/GitHub/datasets/coco8/labels/val.cache... 4 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 4/4 169.5Kit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 1/1 2.5it/s 0.4s
                   all          4         17          0          0          0          0
                person          3         10          0          0          0          0
                   dog          1          1          0          0          0          0
                 horse          1          2          0          0          0          0
              elephant          1          2          0          0          0          0
              umbrella          1          1          0          0          0          0
          potted plant          1          1          0          0          0          0
Speed: 1.1ms preprocess, 43.1ms inference, 0.0ms loss, 23.1ms postprocess per image
💡 Learn more at https://docs.ultralytics.com/modes/val
(base) toxite@macos ultralytics % yolo val model=yolo11n.pt device=mps exist_ok plots=False compile=False
WARNING ⚠️ 'data' argument is missing. Using default 'data=coco8.yaml'.
Ultralytics 8.3.198 🚀 Python-3.12.0 torch-2.6.0 MPS (Apple M4 Pro)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2153.2±1343.6 MB/s, size: 54.0 KB)
val: Scanning /Users/toxite/Downloads/GitHub/datasets/coco8/labels/val.cache... 4 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 4/4 180.4Kit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 1/1 2.5it/s 0.4s
                   all          4         17      0.555      0.833      0.811      0.608
                person          3         10      0.464        0.5      0.371      0.133
                   dog          1          1      0.548          1      0.995      0.697
                 horse          1          2      0.531          1      0.995      0.674
              elephant          1          2      0.371        0.5      0.516      0.256
              umbrella          1          1      0.569          1      0.995      0.995
          potted plant          1          1      0.847          1      0.995      0.895
Speed: 1.1ms preprocess, 41.6ms inference, 0.0ms loss, 23.0ms postprocess per image
💡 Learn more at https://docs.ultralytics.com/modes/val
```
